Add cephadm-signed certificate support for all services

rkachach opened the pull request:
## Summary

This PR builds on top of the `certmgr` (introduced in the PR #61544) to add comprehensive and generic support for both user-provided and cephadm-signed certificates across all supported services. Previously, this support was limited to mgmt-gateway and oauth2-proxy. This enhancement improves default security and simplifies certificate management in cephadm-managed clusters. In addition to enabling cephadm-signed certificates for all services, this PR also introduces a new mechanism for provisioning certificates, using references to certmgr certificate instances.

## Key Enhancements

- **Integration with certmgr**  
Extend certmgr support to include cephadm-signed certificates for all services, automating the entire certificate lifecycle, including automated renewal. For this this PR modifies services/daemons deployment code to automatically use the newly implemented certificate management system, simplifying the code and reducing manual configuration overhead previously needed for each service.

- **Unified Logic for Certificates Management**  
  Unify certificates management logic across all the cephadm supported services. With the changes on this PR we now can add TLS support for new (or legacy) services with fewer changes in the code. 

The certificates mgmt logic in the base class `cephadmservice` will take care of the rest.


- **Simplified Centralized Certificate Management**  
Centralized certificate management helps reduce security risks. This update provides a simple, out-of-the-box solution for automatically getting cephadm-signed certificates for all services, making it easier for users to deploy secure services by using cephadm-signed or provide their own certificates.

## Implementation Details

- **Spec changes** 

The class `ServiceSpec` has been extended to add the following generic new fields:

> - **ssl**: A boolean indicating whether SSL is enabled for the service.
> - **ssl_cert**: The path or reference to the SSL certificate.
> - **ssl_key**: The path or reference to the SSL private key.
> - **certificate_source**: The source of the certificate (see below).

Added `CertificateSource` enum to specify certificate origin:

> - **inline**: cert embedded directly in the spec.
> - **reference**: cert provided via external reference (to certmgr).
> - **cephadm_signed**: cert auto-generated and signed by cephadm (default).
> 

- **Cephadm-signed Certificate Generation** 
Added logic to generate entities for cephadm-signed certificates on the fly

- **How service gets Certificates**  
With the new changes services that support SSL configuration will call the new method `get_certificates` which will take care of all the underlying details of the certificates mgmt such as what certificate source is configured, if the certificates exists or not, etc. With the new code services now have just to call the following method:

```
    def get_certificates(self,
                         daemon_spec: CephadmDaemonDeploySpec,
                         ips: List[str] = [],
                         fqdns: List[str] = [],
                         custom_sans: List[str] = []
                         ) -> Tuple[str, str]:
```

- **Certificates Lifecycle Management**  
Cephadm-signed certificates are created/removed automatically on demand as the services are deployed/removed. User-provided certificates handling depends on the source. For example if the certificate source is `reference` then cephadm will not delete the certificate even if the service has been removed. Similarly in case of `inline` certificates certmgr will not allow the user to modify these directly as the source of truth in this case is the service spec.

## New CLI Commands options for Certificate Management

### `ceph orch certmgr cert ls`
Lists known certificates with optional filtering.

**Options:**
- `--include-cephadm-signed` — Show cephadm-signed certificates in addition to user-provided ones.
- `--filter-by <expression>` — Comma-separated list of filter expressions.
  - Supported fields:
    - `name=<pattern>` — Filter by certificate name (supports wildcards).
    - `scope=<scope>` — One of `global`, `host`, `service`.
    - `service=<pattern>` — Filter by associated service name.
    - `status=<status>` — One of `valid`, `invalid`, `expired`, `expiring`.
    - `signed-by=<source>` — `user` or `cephadm`.

**Examples:**
```bash
ceph orch certmgr cert ls --filter-by "status=expired"
ceph orch certmgr cert ls --filter-by "scope=service,status=expiring"
ceph orch certmgr cert ls --include-cephadm-signed --filter-by "name=rgw*,status=valid"
``` 

## Example Service Specs

### 1. Using `cephadm-signed` (default)
```yaml
service_type: rgw
service_id: rgw
spec:
  ssl: true
  certificate_source: cephadm-signed
```

### 2. Using `reference` 

In this case, we first has to provision the certificate/key by using the `certmgr` commands:

```bash
ceph orch certmgr cert set --cert-name rgw_ssl_cert  --service-name rgw.myrgw -i server_cert
ceph orch certmgr cert set --key-name rgw_ssl_key  --service-name rgw.myrgw -i server_key
```

Once provisioned, then we can use the following spec. Cephadm will automatically get the certificates for this RGW service and use them. 

```yaml
service_type: rgw
service_id: myrgw
spec:
  ssl: true
  certificate_source: cephadm-signed
```

### 2. Using `inline` 

This is the legacy method (embedding the certificates in the spec)

``` yaml
service_type: rgw
service_id: myrgw
spec:
  ssl: true
  certificate_source: inline
  ssl_cert: |
    -----BEGIN CERTIFICATE-----
    MIIDzjCCAragAwIBAgIUFc...
    -----END CERTIFICATE-----
  ssl_key: |
    -----BEGIN PRIVATE KEY-----
    MIIEvQIBADANBgkqhkiG9w0B...
    -----END PRIVATE KEY-----

```
---

### Bindings listing:

In order to get the certificates names, certmgr provides the `cerph orch certmgr bidings ls` command. Following is an example of the output (truncated for simplicity). Those are the names of the certificates that the user has to pass as `--cert-name`/`--key-name` when adding new certificates to the certmgr store.

```
> ceph orch certmgr bindings ls
service:
  rgw:
    certs:
    - rgw_ssl_cert
    keys:
    - rgw_ssl_key
host:
  grafana:
    certs:
    - grafana_ssl_cert
    keys:
    - grafana_ssl_key
global:
  mgmt-gateway:
    certs:
    - mgmt_gateway_ssl_cert
    keys:
    - mgmt_gateway_ssl_key

```

## Testing and Validation

- **Integration Tests**  

- **Upgrade Tests**  




<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

rkachach committed:
mgr/cephadm: adapting iscsi service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting Grafana service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adding support to extract ips and fqdns from cert

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting node-exporter service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: adding new spec generic fields fo ssl/certifiactes

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: adapting mgmt-gateway service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting oauth2-proxy service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: removing call to previous method to storing all certs

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach commented:
jenkins test make check

rkachach committed:
mgr/cepahdm: adapting agent service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting node-proxy service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
doc/cephadm: updating certmgr docs to reflects new changes

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

sayaleeraut commented:
Scenario: Set and remove user-signed certificate for a service 

Observations :

Even if the --service-name is passed as "None", CLI accepts it and gives a success msg-
```
# ceph orch certmgr cert set rgw_ssl_cert -i /etc/ceph/cert.pem --service_name None
Certificate for rgw_ssl_cert set correctly
```

Cert is not removed on passing incorrect service-name, which is expected, but cert rm command gives success msg  -

```
# ceph orch ls --service-type rgw
NAME       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT
rgw.rgw.1  ?:443      2/2  48s ago    2w   label:rgw

# ceph orch certmgr cert rm rgw_ssl_cert
Error EINVAL: Cannot delete the certificate. Please use 'ceph orch certmgr cert ls' to list available certificates.
Note: for certificates with host/service scope use --service-name or --hostname to specify the target.

# ceph orch certmgr cert rm rgw_ssl_cert --service_name rgw
Certificate for rgw_ssl_cert removed correctly

# ceph orch certmgr cert ls
rgw_ssl_cert
  scope: service
  certificates
    None
      subject
        countryName: IN
        stateOrProvinceName: ST
        localityName: XYZ
        organizationName: ORG
        commonName: ceph-saya-upstream-4ezarj-node1-installer
      validity
        remaining_days: 6094
```
        
Tried removing the cert using correct service-name , but it fails -

```
# ceph orch certmgr cert rm rgw_ssl_cert --service_name rgw.rgw.1
Certificate for rgw_ssl_cert removed correctly

# ceph orch certmgr cert ls
rgw_ssl_cert
  scope: service
  certificates
    None
      subject
        countryName: IN
        stateOrProvinceName: ST
        localityName: XYZ
        organizationName: ORG
        commonName: ceph-saya-upstream-4ezarj-node1-installer
      validity
        remaining_days: 6094

# ceph orch certmgr reload
OK

# ceph orch certmgr cert ls
rgw_ssl_cert
  scope: service
  certificates
    None
      subject
        countryName: IN
        stateOrProvinceName: ST
        localityName: XYZ
        organizationName: ORG
        commonName: ceph-saya-upstream-4ezarj-node1-installer
      validity
        remaining_days: 6094
```

rkachach commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -865,6 +902,11 @@ def __init__(self,
                  config: Optional[Dict[str, str]] = None,
+                 ssl: Optional[bool] = False,
+                 certificate_source: Optional[str] = None,
+                 custom_sans: Optional[List[str]] = None,
```
@phlogistonjohn really sorry for the very late response: `custom_sans` are custom SAN entries that the user can configure to insert in the generated certificate. We provide this support for RGW.

rkachach commented:
> I tested the new enhancements, as per discussion with Redouane, the "certificate_source:" should be part of "spec:" of a service when user wants to use it, but by default for all services, it is NOT a part of "spec:" -
> 
> ```
> # ceph orch ls --service-name rgw.rgw.1 --export
> service_type: rgw
> certificate_source: cephadm-signed
> service_id: rgw.1
> service_name: rgw.rgw.1
> placement:
>   label: rgw
> spec:
>   rgw_exit_timeout_secs: 120
> ```
> 
> Please note that this can cause confusion because without "ssl: true", no cephadm-signed cert is generated. So, if a user wants to use cephadm signed cert, then in that case, the service spec needs to be as below (where the "certificate_source:" needs to be moved to "spec:" ) -
> 
> ```
> service_type: rgw
> service_id: rgw.1
> service_name: rgw.rgw.1
> spec:
>   certificate_source: cephadm-signed
>   ssl: true
>   rgw_exit_timeout_secs: 120
> ```
> 
> But, after doing the above, when again the service spec is extracted using --export, the output is as below -
> 
> ```
> # ceph orch ls --service-name rgw.rgw.1 --export 
> service_type: rgw
> certificate_source: cephadm-signed
> ssl: true
> service_id: rgw.1
> service_name: rgw.rgw.1
> placement:
>   label: rgw
> spec:
>   rgw_exit_timeout_secs: 120
> ```
> 
> Why is the arg "certificate_source:" not part of "spec:" in this command output as well ?

Thanks for reporting the issue @sayaleeraut, I fixed the issue and now the field should appear in the `spec` section.

rkachach commented:
> Scenario : Set a cert-key using certmgr command and change the service spec to use the "certificate_source: reference" :
> 
> ```
> # ceph orch certmgr cert-key set grafana -i /var/lib/ceph/good_cert_key.pem --hostname ceph-saya-upstream-4ezarj-node1-installer
> Certificate/key pair set correctly
> ```
> 
> ```
> [ceph: root@ceph-saya-upstream-4ezarj-node1-installer /]# ceph orch certmgr cert ls --include-cephadm-signed
> grafana_ssl_cert
>   scope: host
>   certificates
>     ceph-saya-upstream-4ezarj-node1-installer
>       subject
>         countryName: IN
>         stateOrProvinceName: MH
>         localityName: BLR
>         organizationName: XYZ
>         organizationalUnitName: ABC
>         commonName: DEF
>         1.2.840.113549.1.9.1: abc@xyz.com
>       validity
>         remaining_days: 1094
> cephadm_root_ca_cert
>   scope: global
>   certificates
>     subject
>       commonName: cephadm-root-b5b6092a-6153-11f0-810d-fa163e263848
>     validity
>       remaining_days: 3644
> cephadm-signed_agent_cert
>   scope: host
>   certificates
>     ceph-saya-upstream-4ezarj-node1-installer
>       subject
>         commonName: 10.0.66.219
>       validity
>         remaining_days: 1816
> cephadm-signed_grafana_cert
>   scope: host
>   certificates
>     ceph-saya-upstream-4ezarj-node1-installer
>       subject
>         commonName: 10.0.66.219
>       validity
>         remaining_days: 1086
> cephadm-signed_rgw.rgw.1_cert
>   scope: host
>   certificates
>     ceph-saya-upstream-4ezarj-node2
>       subject
>         commonName: 10.0.67.71
>       validity
>         remaining_days: 1087
> service_type: grafana
>     ceph-saya-upstream-4ezarj-node3
>       subject
>         commonName: 10.0.67.2
>       validity
>         remaining_days: 1087
> ```
> 
> ```
>  # ceph orch ls --service-type grafana --export
> service_type: grafana
> certificate_source: cephadm-signed
> ssl: true
> service_name: grafana
> placement:
>   count: 1
> spec:
>   anonymous_access: true
>   protocol: https
> ```
> 
> To change the spec so that only user added cert is used, create a new spec for Grafana -
> 
> ```
> service_type: grafana
> ssl: true
> service_name: grafana
> placement:
>   count: 1
> spec:
>   certificate_source: reference
>   anonymous_access: true
>   protocol: https
> 
> # ceph orch apply -i /var/lib/ceph/new_grafana.yaml
> Error EINVAL: Need host to access cert for entity grafana_ssl_cert
> ```
> 
> Created again a new spec for grafana -
> 
> ```
> # cat /var/lib/ceph/new_grafana.yaml
> service_type: grafana
> ssl: true
> service_name: grafana
> placement:
>   count: 1
>   hosts:
>     - ceph-saya-upstream-4ezarj-node1-installer
> spec:
>   certificate_source: reference
>   host: ceph-saya-upstream-4ezarj-node1-installer
>   anonymous_access: true
>   protocol: https
> 
> # ceph orch apply -i /var/lib/ceph/new_grafana.yaml
> Error EINVAL: ServiceSpec: __init__() got an unexpected keyword argument 'host'
> ```
> 
> Hey @rkachach Is this expected?

Hi @sayaleeraut 

No, this is not expected. The check was assuming `service-scope` type of services but Grafana is a different case (host-scope). I fixed the code so we don't check for certs existence in case of `host-scoped` services. Internally, even if we would like to, we can't as we don't know on which hosts service daemons will be deployed till very late in the deployment process (this depends on the placement and there're different types of placement). To help/remind users that they need to create the certificates I added a warning to be shown in this case. So you will see a message like:

```
[ceph: root@ceph-node-0 /]# ceph orch apply -i grafana.spec 
Scheduled grafana update... 
Warning: SSL is configured with 'reference', and this service uses per-host certificates.
To configure keys/certificates, run the following commands for each host daemons are deployed on:
  > ceph orch certmgr cert set --cert-name grafana_ssl_cert --service-name grafana --hostname <host>  -i <cert-file>
  > ceph orch certmgr key set --key-name grafana_ssl_cert --service-name grafana --hostname <host> -i <key-file>
Once all certificates are provisioned, run:
  > ceph orch reconfig grafana
to apply the updated certificates.
```

rkachach committed:
mgr/cephadm: fixing nvmeof section in cert_mgr UT + new UT

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach commented:
jenkins test rook e2e

rkachach commented:
jenkins test api

rkachach commented:
jenkins test api

rkachach committed:
mgr/cepahdm: add support for custom duration when generating certs

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: adding self-signed certifiactes support for nvmeof svc

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adding new filtring options and certificate ref checks

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: adding enahanced support for self-signed certs

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: adding support to prevent editing readonly certs

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting UT after all the changes

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: adapting python-common UT to the spec changes

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach commented:
jenkins test make check

anthonyeleven commented on doc/cephadm/certmgr.rst:
```diff
@@ -16,27 +16,24 @@ while offering flexibility for deployment preferences.
 
-This proactive approach ensures secure and uninterrupted service operations
-while allowing users to manage certificate policies according to their needs.
+In this case we distinguish between two different options:
```
Nits: suggest:

`There are two types:`

anthonyeleven commented on doc/cephadm/certmgr.rst:
```diff
@@ -76,24 +73,34 @@ certificate lifecycle and renewal:
-replace certificates before service disruptions occur.
+Cephadm continuously monitors the status of all managed certificates—both self-signed and user-provided.
 
+- If a certificate is invalid or has already expired, Cephadm issues a health
```
Suggest:

all managed certificates, both

anthonyeleven commented on doc/cephadm/certmgr.rst:
```diff
@@ -76,24 +73,34 @@ certificate lifecycle and renewal:
 
+- If a certificate is invalid or has already expired, Cephadm issues a health
+  error (``CEPHADM_CERT_ERROR``) to alert administrators.
+- If the certificate is approaching its expiration date, as determined by the
```
Suggest

if a certificate

anthonyeleven reviewed (commented):
Several minor docs suggestions.  

Also suggest for the PR title, imperative:

`Add cephadm-signed certificate support for all services`

rkachach committed:
mgr/cephadm: adding automated certificates mgmt for cephadmservice

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting alertmanager service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cepahdm: adapting Prometheus service to use the new cert mgmt

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

phlogistonjohn reviewed (commented):
partially done

phlogistonjohn reviewed (commented):
I see no blockers but plenty of opportunities for cleaning and tightening up in the future, so feel free to go ahead with this as-is and tackle my suggestions later.

I do think we (you and I) need to sync up again on cert stuff since I was off doing my own thing WRT certificates in the smb module. Even if it's never fully integrated we ought to be able to harmonize some parts of these two subsystems.

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -16,27 +16,24 @@ while offering flexibility for deployment preferences.
 
-This proactive approach ensures secure and uninterrupted service operations
-while allowing users to manage certificate policies according to their needs.
+There are two types:
+- Inline: Certificates embedded in the service spec; cannot be modified directly via certmgr.
```
An empty line is needed here before the unordered list (between lines 32-33) otherwise the list is rendered as a single paragraph.
```suggestion
There are two types:

- Inline: Certificates embedded in the service spec; cannot be modified directly via certmgr.
```

Also in this list the 2 occurrences of "certmgr" is rendered as normal text while in previous text it was rendered as inline code using double backticks, is that intentional?

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -16,27 +16,24 @@ while offering flexibility for deployment preferences.
+- Inline: Certificates embedded in the service spec; cannot be modified directly via certmgr.
+- Reference: Certificates set and managed via certmgr; must be manually uploaded.
 
+See ``Certificate Health Monitoring`` for details on how certmgr handles user-provided certificate expiration.
```
Suggest a link here instead of inline code.

```suggestion
See `Certificate Health Monitoring`_ for details on how certmgr handles user-provided certificate expiration.
```

Also another occurrence of certmgr without inline code formatting.

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -103,54 +111,54 @@ Cephadm certmgr supports three different scopes for certificate management:
    - Certificates in this scope are shared across all service daemons, regardless of which host they are running on.
-   - Example: `mgmt-gateway` certificate is a globally shared certificate used by all service daemons.
+   - Example: ``mgmt-gateway`` certificate is a globally shared certificate used by all service daemons.
 
 2. **Per-Host Scope:**
    - Certificates are assigned per host, meaning each host has its own unique certificate.
    - When configuring a custom certificate, the user must specify the host for which the certificate applies.
-   - Example: `grafana` service cerificates are configured at the host level and applies specifically to a single machine.
+   - Example: ``grafana`` service certificates are configured at the host level and apply specifically to a single machine.
 
 3. **Per-Service Scope:**
-   - Certificates are configured per service name, meaning each instance of a service can have its own certificate.
+   - Certificates are configured per service name (each service instance can have its own certificate).
    - When specifying a custom certificate, the user must define the service to which it belongs.
-   - Example: an `rgw` service certificate is assigned specifically and only to an RGW service
+   - Example: an ``rgw`` service certificate is assigned specifically and only to an RGW service
```
An empty line is needed before the unordered sub-lists (between lines 113-114, 116-117, 121-122), right now they are just rendered as single paragraphs and not as unordered lists.

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -103,54 +111,54 @@ Cephadm certmgr supports three different scopes for certificate management:
 
 1. **Global Scope:**
```
Suggest to change these 3 items into an ordered list instead of hardcoding the numbers.

```suggestion
#. **Global Scope:**
```

And similarly to lines 116, 121.

rkachach replied:
Fixed

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -103,54 +111,54 @@ Cephadm certmgr supports three different scopes for certificate management:
+Examples:
 
-  ceph orch certmgr cert ls [--show-details]
+.. code-block:: bash
```
Perhaps use prompt block here like the other commands? I see it used to be that before but was changed -- maybe I'm missing some nuance in this case.

```suggestion
.. prompt:: bash #
```

rkachach replied:
Fixed

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -183,9 +192,8 @@ To retrieve the content of a specific certificate:
-Replace `<certificate-name>` with the actual certificate name from `ceph
-orch certmgr cert ls`. For certificates with host or service scope, include
-the `--hostname` or `--service_name` arguments as needed.
+Use `<service-name>` as listed by `ceph orch certmgr bindings ls`. For certificates with host or service scope,
+include the `--hostname` or `--service_name` arguments as needed.
```
In other places of the document (and docs in general) these are using double backticks instead of single backtick.

```suggestion
Use ``<service-name>`` as listed by ``ceph orch certmgr bindings ls``. For certificates with host or service scope,
include the ``--hostname`` or ``--service_name`` arguments as needed.
```

bluikko commented on doc/cephadm/certmgr.rst:
```diff
@@ -207,14 +215,12 @@ To associate a certificate with a private key:
-key in PEM format. This file should include both the certificate and
-private key concatenated together.
+Use this command to upload or replace an existing certificate/key pair for a certain service.
+Replace `<service-type>` with the actual certificate name from `ceph orch certmgr bindings ls`.
```
Same as above.

```suggestion
Replace ``<service-type>`` with the actual certificate name from ``ceph orch certmgr bindings ls``.
```

Two more occurrences of single backticks used for this purpose in the following sections outside of this PR/in the old version of the file.

bluikko reviewed (commented):
A few of formatting errors leading to unintended rendering + some consistency nits.

rkachach commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -811,6 +825,29 @@ class ServiceSpec(object):
         'smb',
     ]
 
+    # List of all services that can requiere TLS certifiactes
```
fixed

rkachach commented on src/pybind/mgr/cephadm/tlsobject_types.py:
```diff
@@ -2,14 +2,42 @@
 from orchestrator import OrchestratorError
 
 
+def parse_bool(value: Any) -> bool:
```
@phlogistonjohn I did a quick search and I didn't find a similar function, plz can you point me to the function? (if any)

rkachach commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -3429,12 +3444,25 @@ def cert_store_set_cert(
+                context = f"host '{hostname}'"
+            raise OrchestratorError(
+                f"Certificate '{cert_name}' for {context} is not editable (defined as inline in the spec or generated by cephadm)."
+            )
```
Goot catch! fixed :)

rkachach commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -284,6 +309,103 @@ def get_dependencies(cls, mgr: "CephadmOrchestrator",
+    def get_certificates(self,
+                         daemon_spec: CephadmDaemonDeploySpec,
+                         ips: List[str] = [],
+                         fqdns: List[str] = []) -> Tuple[str, str]:
```
@phlogistonjohn you wouldn’t believe how much code I had to touch just to switch this return type to a NamedTuple 😅 ... Totally worth it though, as result the code is much cleaner and simpler now. Thanks!

bluikko reviewed (commented):
Looks good to me, only 1 case that renders incorrectly.

phlogistonjohn reviewed (approved):
thanks for taking so many of my suggestions into account in this PR. While I'm no expert on this topic nothing jumps out at me as an issue. So squash patches at your convenience it all generally lgtm

rkachach committed:
mgr/cephadm: Adding RGW migration for the new certmgr certs format

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach committed:
mgr/cephadm: using 5 years for service-discovery internal certs

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach commented on doc/cephadm/certmgr.rst:
```diff
@@ -16,27 +16,25 @@ while offering flexibility for deployment preferences.
+- Inline: Certificates embedded in the service spec; cannot be modified directly via ``certmgr``.
+- Reference: Certificates set and managed via ``certmgr``; must be manually uploaded.
+
+See ``Certificate Health Monitoring``_ for details on how ``certmgr`` handles user-provided certificate expiration.
```
Fixed

adk3798 commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -1248,20 +1249,15 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
             san_list = spec.zonegroup_hostnames or []
-            custom_san_list = san_list + [f"*.{h}" for h in san_list] if spec.wildcard_enabled else san_list
-
-            cert, key = self.mgr.cert_mgr.generate_cert(
-                daemon_spec.host,
-                self.mgr.inventory.get_addr(daemon_spec.host),
-                custom_san_list=custom_san_list
-            )
```
this was previously generating a certificate per RGW and setting each one up to be for the specific host it was deployed on. I admittedly can't remember why and if that was necessary (tracked it to https://github.com/ceph/ceph/commit/66e9ef16a93fc542983fb637e562278f9dc4847b but I didn't comment much on the host specific part of it). I don't think I have any recommendations or anything, but if we are changing the behavior, just something to remember if we see any auth/cert issues with RGW later.

adk3798 commented on src/pybind/mgr/cephadm/services/nvmeof.py:
```diff
@@ -43,6 +44,64 @@ def config(self, spec: NvmeofServiceSpec) -> None:  # type: ignore
+        """
+
+        if not spec.ssl:
+            self.mgr.log.info("TLS for nvmeof gw is disabled.")
```
we might want to also include the service name in this error, in case users have multiple nvmeof services

rkachach replied:
Fixed

adk3798 commented on src/pybind/mgr/cephadm/migrations.py:
```diff
@@ -81,34 +86,57 @@ def verify_no_migration(self) -> None:
+            logger.info('trying migrate_1_2')
             if self.migrate_1_2():
+                logger.info('migrate_1_2 Done')
                 self.set(2)
 
         if self.mgr.migration_current == 2 and not startup:
+            logger.info('trying migrate_2_3')
             if self.migrate_2_3():
+                logger.info('migrate_2_3 Done')
                 self.set(3)
 
         if self.mgr.migration_current == 3:
+            logger.info('trying migrate_3_4')
             if self.migrate_3_4():
+                logger.info('migrate_3_4 Done')
                 self.set(4)
 
         if self.mgr.migration_current == 4:
+            logger.info('trying migrate_4_5')
             if self.migrate_4_5():
+                logger.info('migrate_4_5 Done')
                 self.set(5)
 
         if self.mgr.migration_current == 5:
+            logger.info('trying migrate_5_6')
             if self.migrate_5_6():
+                logger.info('migrate_5_6 Done')
                 self.set(6)
 
         if self.mgr.migration_current == 6:
+            logger.info('trying migrate_6_7')
             if self.migrate_6_7():
+                logger.info('migrate_6_7 Done')
                 self.set(7)
 
+        if self.mgr.migration_current == 7:
+            logger.info('trying migrate_7_8')
+            if self.migrate_7_8():
+                logger.info('migrate_7_8 Done')
+                self.set(8)
```
given these are all identical other than the number, I'm wondering if there is some way to sort of automate this bit where it automatically loops from the current `migration_current` up to whatever `LAST_MIGRATION` is and handles logging the right numbers and calling the migration functions with the right numbers so we don't need to keep worrying about this. Probably not a task for this PR, but something to think about

adk3798 reviewed (commented):
As usual with the large PRs, I reviewed by commits so if you fixed anything I commented on in a later commit and I didn't notice just mark it resolved.

No specific blockers I see other than the iscsi issue. I'm also generally a bit weary about any potential unintended side effects to the changes around `prepare_create` and `generate_config` mostly in terms of which classes define them and when they're getting called. I have nothing definitive to complain about  on that front (although the iscsi issue seems like it might be related) but it's just something to be careful with.

rkachach commented on src/pybind/mgr/cephadm/services/iscsi.py:
```diff
@@ -52,6 +52,7 @@ def get_dependencies(cls, mgr: "CephadmOrchestrator",
     def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonDeploySpec:
         assert self.TYPE == daemon_spec.daemon_type
 
+        super().prepare_create(daemon_spec)
```
@adk3798 Good catch! thanks. I did a test where I essentially changed the call and the issue disappeared. I fixed the code by removing the call to the `prepare-create` and just called directly the `self.get_certificates(daemon_spec)` which in fact the final goal of calling `prepare-create` of the parent. During the change I notice that the iscsii daemon structure is different (it does everything on the `prepare-create` and doesn't declare a generate_config). Anyway, to prevent regressions I didn't change anything else as iscsi is mainly a legacy service at this point.

rkachach commented on src/pybind/mgr/cephadm/cert_mgr.py:
```diff
@@ -204,23 +293,84 @@ def get_key(self, key_name: str, service_name: Optional[str] = None, host: Optio
+        # with the exception of the cephadm root CA cert (CEPHADM_ROOT_CA_CERT) as
+        # technically the user may be interested in adding it to his CA trust chain
+        filters = [_make_filter(expr) for expr in filter_by.split(",") if expr.strip()]
+        if not include_cephadm_signed and not any('signed-by=' in f for f in filter_by.split(',')):
+            filters.append(lambda ctx: ctx['signed_by'] == 'user' or ctx['name'] == self.CEPHADM_ROOT_CA_CERT)
```
Well, I think it's not duplicate logic, there are two layers with clear precedence: 1) _make_filter applies any explicit `signed-by=...` from `filter_by`. The extra check is just a default that kicks in only when there’s no `signed-by` filter and `include_cephadm_signed=False`, so we show `user-made` certs and the root CA by default. If you set `include_cephadm_signed=True` or specify `signed-by=...`, the default is skipped and we apply the explicit filter from the user (whatever he has put in the filter-by). I've add a doc string to explain the behavior.

rkachach commented on src/pybind/mgr/cephadm/cert_mgr.py:
```diff
@@ -231,10 +381,12 @@ def cert_ls(self, include_datails: bool = False) -> Dict:
         key_objects: List = self.key_store.list_tlsobjects()
         ls: Dict = {}
         for key_name, key_obj, target in key_objects:
+            if not include_cephadm_signed and self.is_cephadm_signed_object(key_name):
```
yeah, it's a bit confusing. I changed the name of the variable to `include_cephadm_generated_keys` to be more explicit.

rkachach commented on src/pybind/mgr/cephadm/cert_mgr.py:
```diff
@@ -204,23 +293,84 @@ def get_key(self, key_name: str, service_name: Optional[str] = None, host: Optio
+
+            if key == 'name':
+                return lambda ctx: fnmatch(ctx['name'], value)
+            if key == 'scope':
+                want = value.upper()
+                return lambda ctx: ctx['scope'].name == want
+            if key == 'service':
+                return lambda ctx: fnmatch(ctx['service'], value)
+            if key == 'status':
+                want = value.lower()
+                return lambda ctx: ctx['status'] == want
+            if key == 'signed-by':
+                return lambda ctx: ctx['signed_by'] == value.lower()
```
I worked out the whole section to make it clear. I introduced a new Enum type for the options.

rkachach commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -723,23 +722,13 @@ def _init_cert_mgr(self) -> None:
-        self.cert_mgr.register_cert_key_pair('oauth2-proxy', 'oauth2_proxy_cert', 'oauth2_proxy_key', TLSObjectScope.GLOBAL)
+        for svc in service_registry.get_all_services():
+            if svc.allows_user_certificates:
+                assert svc.SCOPE != TLSObjectScope.UNKNOWN, f"Service {svc.TYPE} requieres certificates but it has not defined its svc.SCOPE field."
```
Fixed

rkachach commented on src/pybind/mgr/cephadm/tlsobject_store.py:
```diff
@@ -1,160 +1,243 @@
-            return (None, None)
+            return TLSObjectTarget(service=None, host=None)
+
+    def get_tlsobject_scope_and_target(self, obj_name: str,
```
Yeah, I know that can be a little bit confusing. So basically in addition to the scope, every TLS object may have an associated target. A target is the concrete identifier within a given scope, for example:

   - For **HOST-scoped objects**: the target is the host FQDN or inventory-ip name where the certificate is bound.
   - For **SERVICE-scoped objects**: the target is the service instance name (e.g. "rgw.myzone").
   - For **GLOBAL objects**, the target is always None because the object applies cluster-wide.

Scopes define ***where*** at what level in the system a TLS object conceptually lives (kind of the namespace), while targets specify the ***which one*** within that scope. The TLSObjectStore uses both pieces of information when saving, retrieving, or removing certificates/keys, ensuring that a single logical object name can have
multiple per-target instances when required.

So in summary the **Scope acts like a namespace**: it partitions all cert/key names into three disjoint “buckets” **GLOBAL, HOST, SERVICE**. Within each namespace, you may have multiple instances identified by a target (host, service instance, or none for global).

Examples:

- **cephadm_root_ca_cert** or **mgmt-gateway**: scope=GLOBAL, target=None
- **rgw_ssl_cert** for rgw.myzone: scope=SERVICE, target="rgw.myzone"
- **cephadm-signed_grafana_cer**t on host node12:  scope=HOST, target="node12"

I'll add this info to the CertMgr docstring for future references.

rkachach commented on src/pybind/mgr/cephadm/tlsobject_store.py:
```diff
@@ -1,160 +1,243 @@
+            except json.JSONDecodeError as e:
+                logger.warning(
+                    f"TLSObjectStore: Cannot parse JSON for '{obj_name}': "
+                    f"key={k}, len={len(v) if v else 0}, startswith={v[:20]!r}, error={e}"
```
@adk3798 no, actually is not. This just a simple way to log the beginning of what's stored there. Since now we need these entries to be JSON, if some user (by mistake) store the raw certificate/key or anything else there we can see what's actually stored there. I only log the first characters bcz the certs can be very long and because of security reasons bcz I don't want to leak secrets (specially private keys) by mistake in the logs.

rkachach commented on doc/cephadm/certmgr.rst:
```diff
@@ -76,81 +74,95 @@ certificate lifecycle and renewal:
+   ceph orch certmgr bindings ls
 
-Entities represent services that utilize managed certificates.
+List services that utilize managed certificates.
```
It applies to all the certificates. I changed the sentence a bit to make it more generic. Plz let me know if you have suggestions on the wording.

rkachach commented:
jenkins test rook e2e

rkachach commented:
jenkins test windows

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
jenkins test api

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
jenkins test dashboard cephadm

rkachach committed:
mgr/cephadm: adapting ingress svc to use the new cert mgmt approach

Signed-off-by: Redouane Kachach <rkachach@ibm.com>

rkachach commented:
jenkins test docs

rkachach commented:
jenkins test windows

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
jenkins test api

rkachach commented:
jenkins test dashboard cephadm

rkachach commented:
Teuthology run results: https://pulpito.ceph.com/saraut-2025-09-04_05:04:51-orch:cephadm-wip-saraut08-testing-2025-09-03-2042-distro-default-smithi/

mitch000001 commented:
Are there any plans that this gets backported/included in tentacle?

rkachach commented:
@mitch000001 Since umbrella has already been kicked off, not sure if this will be backported to Tentacle. However certmgr is already present in tentacle but with limited functionality.
